src/HexagonOptimize.cpp

```diff
@@ -860,14 +885,77 @@ class OptimizePatterns : public IRMutator2 {
         if (op->type.is_vector()) {
             Expr cast = op;
 
+            // Truncating multiplies require special care, because the
+            // simplifier can cause them to have denominators we do not expect.
+            // If the simplifier cancels a factor out of these patterns, we can
+            // still use them, but we have to inject the factor back into the
+            // expression.
+            vector<Expr> matches;
+            for (const Pattern &p : trunc_mpy) {
+                if (!check_pattern_target(p.flags, target)) {
+                    continue;
+                }
+
+                if (expr_match(p.pattern, cast, matches)) {
+                    int log2_denominator = 0;
+                    if (matches.size() == 4) {
+                        // Rounding patterns have 4 operands, with the rounding
+                        // in the 3rd operand and the denominator in the 4th.
+                        if (!is_const_power_of_two_integer(matches[3], &log2_denominator)) {
+                            continue;
+                        }
+                        if (!can_prove(matches[2] * 2 == i64(1) << log2_denominator)) {
+                            continue;
+                        }
+                    } else {
+                        // The non-rounding patterns have the denominator in the 3rd operand.
+                        if (!is_const_power_of_two_integer(matches[2], &log2_denominator)) {
+                            continue;
+                        }
+                    }
+                    // Drop the divisor and the rounding.
+                    matches.resize(2);
+
+                    // If the power of 2 is not exactly 2^(bits of the result
+                    // type), we need to scale up the operand accordingly.
+                    int shift = cast.type().bits() - log2_denominator;
+                    if (p.intrin.find("mpy2") != std::string::npos) {
+                        // Account for a built-in factor of 2.
+                        shift -= 1;
+                    }
+
+                    // We need to build the scale into one of the operands,
+                    // which must be a constant.
+                    if (is_const(matches[0])) {
```

Comment:
You could perhaps move this until after calling process_match_flags?That function will hand SwapOps01. You can then scale up matches[1] always.

Reply:
I want to do this before, because process_match_flags will strip away widening casts. I want to apply the scale before stripping the widening casts, because the simplifier does the right thing (i32(x)*2 doesn't simplify, while i32(10)*2 does).

Will add a comment.